### ISSUE TITLE
[Serve] Connection refused when issuing requests to remote deployment

### ISSUE THREAD

**Issue by erilju719** [ray-project/ray#28324]
### What happened + What you expected to happen

I can successfully run my deployment on a remote machine (Ubuntu18 desktop)
```
(ServeController pid=9612) INFO 2022-09-07 09:13:15,438 controller 9612 http_state.py:129 - Starting HTTP proxy with name 'SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-448361050691a2c28bb93f60d6a6f81749fb9a1e10b875300d179b82' on node '448361050691a2c28bb93f60d6a6f81749fb9a1e10b875300d179b82' listening on '127.0.0.1:8000'
(HTTPProxyActor pid=9644) INFO:     Started server process [9644]
(ServeController pid=9612) INFO 2022-09-07 09:13:16,588 controller 9612 deployment_state.py:1232 - Adding 1 replicas to deployment 'unpack_request'.
(ServeController pid=9612) INFO 2022-09-07 09:13:16,619 controller 9612 deployment_state.py:1232 - Adding 1 replicas to deployment 'capitalize'.
(ServeController pid=9612) INFO 2022-09-07 09:13:16,639 controller 9612 deployment_state.py:1232 - Adding 1 replicas to deployment 'capitalize_1'.
(ServeController pid=9612) INFO 2022-09-07 09:13:16,659 controller 9612 deployment_state.py:1232 - Adding 1 replicas to deployment 'create_full_name'.
(ServeController pid=9612) INFO 2022-09-07 09:13:16,671 controller 9612 deployment_state.py:1232 - Adding 1 replicas to deployment 'DAGDriver'.
2022-09-07 09:13:18,644 SUCC scripts.py:307 -- Deployed successfully.
```
but my requests are refused by it (`first_name` and `last_name` are the two inputs to the graph):
```
C:\Users\eljl>curl -X PUT "http://<remote_machine_ip>:8000/?first_name=a&last_name=b"
curl: (7) Failed to connect to <remote_machine_ip> port 8000 after 2063 ms: Connection refused
```
I have verified that it is possible to communicate over port 8000 from my local Windows laptop to the remote machine by first issuing `nc -l 8000` on the remote machine, and then run `telnet <remote_machine_ip> 8000`. With that, I can send strings back and forth. I also verified that `PingSucceeded: True` with the `Test-NetConnection` Powershell cmdlet.

### Versions / Dependencies

Local machine: Windows 10, Python 3.9.13, `ray==2.0.0`
Remote machine: Ubuntu 18, Python 3.9.13, `ray==2.0.0`

### Reproduction script
```
from ray import serve
from ray.serve.drivers import DAGDriver
from ray.serve.deployment_graph import InputNode
from starlette.requests import Request


@serve.deployment
def unpack_request(http_request: Request) -> tuple[str, str]:
    return http_request.query_params.get("first_name"), http_request.query_params.get("last_name")


@serve.deployment
def capitalize(name: tuple[str, str], index: int) -> str:
    return name[index].capitalize()


@serve.deployment
def create_full_name(first_name: str, last_name: str):
    return " ".join((first_name, last_name))


with InputNode() as input:
    name = unpack_request.bind(input)
    first_name_cap = capitalize.bind(name, index=0)
    last_name_cap = capitalize.bind(name, index=1)
    full_name = create_full_name.bind(first_name_cap, last_name_cap)

graph = DAGDriver.bind(full_name)
```
### Issue Severity

High: It blocks me from completing my task.

**Comment by sihanwang41** [comment]
Hi @erilju719 , have you used the `serve.run(graph)` after the `DAGDriver.bind(full_name)`?  if yes, could you paste the completed whole log?

**Comment by shrekris-anyscale** [comment]
I ran the repro on a local Ray cluster, and I tried running the `curl` command to send a request to it locally. I noticed two issues with the repro that may/may not be related to the underlying issue:

1. **(Possible related)** I got a `Method Not Allowed` error with the `curl` command from the repro:

```console
(ae) shrekris@Shreyas-MacBook-Pro scratch % curl -X PUT "http://localhost:8000/?first_name=a&last_name=b"
{"detail":"Method Not Allowed"}%   
```

Changing the `curl` request to a `GET` method worked locally:

```console
(ae) shrekris@Shreyas-MacBook-Pro scratch % curl -X GET "http://localhost:8000/?first_name=a&last_name=b"
"A B"%    
```

2. **(Likely unrelated)** The `tuple[x, y]` type annotation isn't correct Python syntax. I had to add

```python
from typing import Tuple
```

and change lines 9 and 14 of the repro to use `Tuple` in the type hint instead of `tuple`. Here's the repro I got up running:

```python
from ray import serve
from ray.serve.drivers import DAGDriver
from ray.serve.deployment_graph import InputNode
from starlette.requests import Request
from typing import Tuple


@serve.deployment
def unpack_request(http_request: Request) -> Tuple[str, str]:
    return http_request.query_params.get("first_name"), http_request.query_params.get("last_name")


@serve.deployment
def capitalize(name: Tuple[str, str], index: int) -> str:
    return name[index].capitalize()


@serve.deployment
def create_full_name(first_name: str, last_name: str):
    return " ".join((first_name, last_name))


with InputNode() as input:
    name = unpack_request.bind(input)
    first_name_cap = capitalize.bind(name, index=0)
    last_name_cap = capitalize.bind(name, index=1)
    full_name = create_full_name.bind(first_name_cap, last_name_cap)

graph = DAGDriver.bind(full_name)
```

**Comment by erilju719** [comment]
@shrekris-anyscale since Python 3.9 you can use the built-in types for type hinting instead of the capitalized types from `Typing`: https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections . So you must be using a different Python version.

**Comment by shrekris-anyscale** [comment]
> since Python 3.9 you can use the built-in types for type hinting instead of the capitalized types from Typing

Ah thanks @erilju719, that makes sense. I'm running Python 3.8.

> Changing the curl request to a GET method worked locally

Did changing the request method have any effect?

**Comment by erilju719** [comment]
> Did changing the request method have any effect?
No, I'm afraid not. Connection refused for both GET and PUT. The reason I used PUT was because that is used on https://docs.ray.io/en/latest/serve/dev-workflow.html.

**Comment by erilju719** [comment]
Turns out the issue was that due to https://github.com/ray-project/ray/issues/28312, any custom `host` and `port` was ignored, so I could not make it listen to any other hosts but 127.0.0.1. To be able to set `--host=0.0.0.0` I had to use the Serve Python API (`serve.run(...)`) and run the script on the remote machine itself. Then, I could successfully issue requests from my Windows laptop. This behaviour is not consistent with the docs (https://docs.ray.io/en/latest/serve/dev-workflow.html#testing-on-a-remote-cluster) though.
